crates/miden-objects/src/address/mod.rs

```diff
@@ -1,364 +1,230 @@
-    /// - The tag length is not [`NoteTag::DEFAULT_NETWORK_TAG_LENGTH`] for network accounts.
-    pub fn with_tag_len(mut self, tag_len: u8) -> Result<Self, AddressError> {
-        if self.id.storage_mode() == AccountStorageMode::Network {
-            if tag_len != NoteTag::DEFAULT_NETWORK_TAG_LENGTH {
-                return Err(AddressError::CustomTagLengthNotAllowedForNetworkAccounts(tag_len));
+    /// - The tag length routing parameter is not [`NoteTag::DEFAULT_NETWORK_TAG_LENGTH`] for
+    ///   network accounts.
+    pub fn with_routing_parameters(
+        mut self,
+        routing_params: RoutingParameters,
+    ) -> Result<Self, AddressError> {
+        if let Some(tag_len) = routing_params.note_tag_len() {
+            match self.id {
+                AddressId::AccountId(account_id) => {
+                    if account_id.storage_mode() == AccountStorageMode::Network
+                        && tag_len != NoteTag::DEFAULT_NETWORK_TAG_LENGTH
+                    {
+                        return Err(AddressError::CustomTagLengthNotAllowedForNetworkAccounts(
+                            tag_len,
+                        ));
+                    }
+                },
             }
-        } else if tag_len > NoteTag::MAX_LOCAL_TAG_LENGTH {
-            return Err(AddressError::TagLengthTooLarge(tag_len));
         }
 
-        self.tag_len = tag_len;
+        self.routing_params = Some(routing_params);
+
         Ok(self)
     }
 
-    // PUBLIC ACCESSORS
+    // ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns the underlying account id.
-    pub fn id(&self) -> AccountId {
+    /// Returns the identifier of the address.
+    pub fn id(&self) -> AddressId {
         self.id
     }
 
+    /// Returns the [`AddressInterface`] of the account to which the address points.
+    pub fn interface(&self) -> Option<AddressInterface> {
+        self.routing_params.as_ref().map(RoutingParameters::interface)
+    }
+
     /// Returns the preferred tag length.
     ///
     /// This is guaranteed to be in range `0..=30` (e.g. the maximum of
     /// [`NoteTag::MAX_LOCAL_TAG_LENGTH`] and [`NoteTag::DEFAULT_NETWORK_TAG_LENGTH`]).
     pub fn note_tag_len(&self) -> u8 {
-        self.tag_len
-    }
-
-    /// Returns the [`AddressInterface`] of the account to which the address points.
-    pub fn interface(&self) -> AddressInterface {
-        self.interface
+        match self.id {
+            AddressId::AccountId(id) => self
+                .routing_params
+                .as_ref()
+                .and_then(RoutingParameters::note_tag_len)
+                .unwrap_or_else(|| {
+                    if id.storage_mode() == AccountStorageMode::Network {
+                        NoteTag::DEFAULT_NETWORK_TAG_LENGTH
+                    } else {
+                        NoteTag::DEFAULT_LOCAL_TAG_LENGTH
+                    }
+                }),
```

Comment:
I think I see now why it makes sense the tag optional in the `RoutingParameters`: the default tag length may depends on the specifics of the identifier. In this case, it is probably fine to keep the current approach (i.g., please ignore that comment).

One thing I'd potentially change is to move default tag length logic into the `AddressId` struct. For example, it would be nice to have `AddressId::default_note_tag_len()` and make the code here much simpler.

Reply:
> the default tag length may depends on the specifics of the identifier

Yeah, that's pretty much it. 

Making it non-optional means we force the sender to determine a default value by either providing one or running the default logic based on the address ID. This is conceptually cleaner, but will probably be awkward when constructing `RoutingParameters`, as one would already need to know the `AddressId`, and the current approach delays this a bit more nicely.